### PR TITLE
GLSP-102: Move tools and their feedback into dedicated modules

### PR DESCRIPTION
**Pull request by martin-fleck-at**
- Restructure code to separate tools module into dedicated modules
- Add custom methods to mouse and key tool to automate de-registration
- Return disposable from feedback dispatcher to return de-registration
- Extract parent class for creation tools
- Use disposable collections for cleanup in parent classes where useful

Fixes https://github.com/eclipse-glsp/glsp/issues/102

### CODE REVIEW

File: packages/client/src/features/tools/change-bounds/change-bounds-tool.ts [hunk 1]
```diff
@@ -183,7 +167,7 @@ export class ChangeBoundsListener extends DragAwareMouseListener implements ISel
                 const resizeActions = this.handleResizeOnClient(positionUpdate);
                 actions.push(...resizeActions);
             }
-            this.tool.dispatchFeedback(actions, this);
+            return actions;
```
Comment (planger) [inline packages/client/src/features/tools/change-bounds/change-bounds-tool.ts:170]: Doesn't this undo the [earlier fix](https://github.com/eclipse-glsp/glsp-client/pull/262) showing moves as feedback (and thus re-apply on `UpdateModelAction`?

File: packages/client/src/container-modules.ts [hunk 2]
```diff
@@ -42,12 +42,19 @@ import modelHintsModule from './features/hints/di.config';
...
 import sourceModelWatcherModule from './features/source-model-watcher/di.config';
 import svgMetadataModule from './features/svg-metadata/di.config';
 import toolFeedbackModule from './features/tool-feedback/di.config';
 import toolPaletteModule from './features/tool-palette/di.config';
-import { enableDefaultToolsOnFocusLossModule, toolsModule } from './features/tools/di.config';
+import changeBoundsToolModule from './features/tools/change-bounds/di.config';
+import deletionToolModule from './features/tools/deletion/di.config';
+import edgeCreationToolModule from './features/tools/edge-creation/di.config';
+import edgeEditToolModule from './features/tools/edge-edit/di.config';
+import { enableDefaultToolsOnFocusLossModule } from './features/tools/enable-default-tools-on-focus-loss';
```
Comment (tortmayr) [inline packages/client/src/container-modules.ts:55]: Unrelated, but maybe you could fix it while you are at it:
Looks like the `enable-default-tools-on-focus-loss` module is not default exported (in contrast to all other modules)

File: packages/client/src/features/tools/edge-creation/edge-creation-tool-feedback.ts [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+            .reverse()
+            .find(element => isConnectable(element) && element.canConnect(edge, 'target'));
+
+        if (endAtMousePosition instanceof SConnectableElement && edge.source && isBoundsAware(edge.source)) {
+            const anchor = this.computeAbsoluteAnchor(endAtMousePosition, Bounds.center(toAbsoluteBounds(edge.source)));
+            if (Point.euclideanDistance(anchor, edgeEnd.position) > 1) {
+                return [MoveAction.create([{ elementId: edgeEnd.id, toPosition: anchor }], { animate: false })];
+            }
+        } else {
+            return [MoveAction.create([{ elementId: edgeEnd.id, toPosition: position }], { animate: false })];
```
Comment (planger) [inline packages/client/src/features/tools/edge-creation/edge-creation-tool-feedback.ts:58]: Shouldn't this be registered as a feedback instead of directly returned?

File: packages/client/src/features/tools/edge-creation/edge-creation-tool-feedback.ts [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        const edge = edgeEnd.feedbackEdge;
+        const position = getAbsolutePosition(edgeEnd, event);
+        const endAtMousePosition = findChildrenAtPosition(target.root, position)
+            .reverse()
+            .find(element => isConnectable(element) && element.canConnect(edge, 'target'));
+
+        if (endAtMousePosition instanceof SConnectableElement && edge.source && isBoundsAware(edge.source)) {
+            const anchor = this.computeAbsoluteAnchor(endAtMousePosition, Bounds.center(toAbsoluteBounds(edge.source)));
+            if (Point.euclideanDistance(anchor, edgeEnd.position) > 1) {
+                return [MoveAction.create([{ elementId: edgeEnd.id, toPosition: anchor }], { animate: false })];
```
Comment (planger) [inline packages/client/src/features/tools/edge-creation/edge-creation-tool-feedback.ts:55]: Shouldn't this be registered as a feedback instead of directly returned?

File: packages/client/src/features/edit-label/edit-label-tool.ts [hunk 5]
```diff
@@ -37,14 +34,7 @@ export class DirectLabelEditTool extends BaseGLSPTool {
...
     enable(): void {
-        this.editLabelKeyListener = this.createEditLabelKeyListener();
-        this.editLabelMouseListener = this.createEditLabelMouseListener();
-        this.mouseTool.register(this.editLabelMouseListener);
-        this.keyTool.register(this.editLabelKeyListener);
-    }
-
-    disable(): void {
-        this.keyTool.deregister(this.editLabelKeyListener);
-        this.mouseTool.deregister(this.editLabelMouseListener);
```
Comment (tortmayr) [inline packages/client/src/features/edit-label/edit-label-tool.ts:48]: Clean refactor of the BaseGLSPTools. Love it!

File: packages/client/src/features/routing/di.config.ts [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+import { ContainerModule } from 'inversify';
+import { ManhattanEdgeRouter, bindOrRebind } from '~glsp-sprotty';
+import { GLSPManhattanEdgeRouter } from './glsp-manhattan-edge-router';
+
+/**
+ * Registers the default tools of GLSP (node and edge creation, changing bounds, edge editing, deletion)
```
Comment (tortmayr) [inline packages/client/src/features/routing/di.config.ts:21]: This comment looks like a copy&paste error

**Review by planger — COMMENTED** [review]
I like this a lot, thank you very much!
I just had a few questions above inline, once I understand those, I'll do a bit more testing.
Thanks again!

File: packages/client/src/features/tool-feedback/di.config.ts [hunk 7]
```diff
@@ -14,48 +14,13 @@
...
-    DrawFeedbackEdgeSourceCommand,
-    HideEdgeReconnectHandlesFeedbackCommand,
-    ShowEdgeReconnectHandlesFeedbackCommand,
-    SwitchRoutingModeCommand
-} from './edge-edit-tool-feedback';
-import { DrawMarqueeCommand, RemoveMarqueeCommand } from './marquee-tool-feedback';
-import { FeedbackEdgeEndView, SResizeHandleView } from './view';
+import { LocationPostprocessor, MoveCommand, TYPES, bindAsService, configureCommand } from '~glsp-sprotty';
 
 const toolFeedbackModule = new ContainerModule((bind, _unbind, isBound) => {
```
Comment (tortmayr) [inline packages/client/src/features/tool-feedback/di.config.ts:19]: Shouldn't t his be part of the `base/feedback`? Since other modules also require the functionality it should not be part of an optional feature module.

File: packages/client/src/features/tools/base-glsp-tool.ts [hunk 8]
```diff
@@ -40,11 +47,11 @@ export abstract class BaseGLSPTool implements GLSPTool {
...
     }
 
-    dispatchFeedback(actions: Action[], feedbackEmitter?: IFeedbackEmitter): void {
-        this.feedbackDispatcher.registerFeedback(feedbackEmitter ?? this, actions);
+    registerFeedback(actions: Action[], feedbackEmitter?: IFeedbackEmitter, cleanupActions?: Action[]): Disposable {
+        return this.feedbackDispatcher.registerFeedback(feedbackEmitter ?? this, actions, cleanupActions);
     }
 
-    deregisterFeedback(actions: Action[], feedbackEmitter?: IFeedbackEmitter): void {
+    deregisterFeedback(actions?: Action[], feedbackEmitter?: IFeedbackEmitter): void {
```
Comment (tortmayr) [inline packages/client/src/features/tools/base-glsp-tool.ts:54]: Parameter order should be reversed. Otherwise you have to call the method with undefined to deregister an emitter without actions. (`deregiserFeedback(undefined, someEmitter)`)
While we are at it we should also rename the `actions` parameter to `cleanupActions`

File: packages/client/src/features/tools/change-bounds/change-bounds-tool-feedback.ts [hunk 9]
```diff
@@ -280,10 +280,12 @@ export class FeedbackMoveMouseListener extends MouseListener implements Disposab
             if (moveAction) {
                 result.push(moveAction);
             }
+            const feebackCleanup: Action[] = [];
```
Comment (tortmayr) [inline packages/client/src/features/tools/change-bounds/change-bounds-tool-feedback.ts:283]:
```suggestion
            const feedbackCleanup: Action[] = [];
```
Although getting a fee back also sound good 😉

File: packages/client/src/features/tools/change-bounds/change-bounds-tool.ts [hunk 10]
```diff
@@ -326,23 +310,9 @@ export class ChangeBoundsListener extends DragAwareMouseListener implements ISel
...
-                    {
-                        elementId: this.activeResizeElement.id,
-                        newPosition: this.initialBounds,
-                        newSize: this.initialBounds
-                    }
-                ]);
-                this.tool.deregisterFeedback([resetResizeAction], this);
-            }
-            this.tool.deregisterFeedback([HideChangeBoundsToolResizeFeedbackAction.create()], this);
+            this.tool.registerFeedback([HideChangeBoundsToolResizeFeedbackAction.create()], this);
```
Comment (tortmayr) [inline packages/client/src/features/tools/change-bounds/change-bounds-tool.ts:315]: Shouldn't this be a `deregisterFeedback`?

File: packages/client/src/features/tools/change-bounds/change-bounds-tool.ts (hunk 1 above)
Comment (tortmayr) [inline packages/client/src/features/tools/change-bounds/change-bounds-tool.ts:170]: This does revert the change of https://github.com/eclipse-glsp/glsp-client/pull/262.
The idea was to dispatch the bounds changes as feedback so that they don't get reset if you receive a server update while resizing.

**Review by tortmayr — CHANGES_REQUESTED** [review]
Awesome change and clean refactoring 😍 
I only had a code review for now and have some remarks.
Will do a complete test later today

File: packages/client/src/features/tools/change-bounds/change-bounds-tool.ts (hunk 1 above)
Comment (martin-fleck-at) [reply packages/client/src/features/tools/change-bounds/change-bounds-tool.ts:170]: Maybe I merged something incorrectly, will double check! I tested some aspects but maybe not all.

File: packages/client/src/base/view/glsp-mouse-tool.ts [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+import { injectable } from 'inversify';
+import { Disposable, MouseListener, MouseTool } from '~glsp-sprotty';
+
+@injectable()
+export class GLSPMouseTool extends MouseTool {
```
[the remark this answers is not in the document]
Comment (martin-fleck-at) [reply packages/client/src/base/view/glsp-mouse-tool.ts:20]: Understood, I'll combine them into a single tool.

File: packages/client/src/base/feedback/feedback-action-dispatcher.ts [hunk 12]
```diff
@@ -35,21 +35,22 @@ export interface IFeedbackActionDispatcher {
     /**
      * Registers `actions` to be sent out by a `feedbackEmitter`.
      * @param feedbackEmitter the emitter sending out feedback actions.
-     * @param actions the actions to be sent out.
+     * @param feedbackActions the actions to be sent out.
+     * @returns A 'Disposable' that de-registers the `feedbackEmitter` with the given `cleanupActions`.
```
[the remark this answers is not in the document]
Comment (martin-fleck-at) [reply packages/client/src/base/feedback/feedback-action-dispatcher.ts:39]: Good catch, thank you!

File: packages/client/src/features/tools/base-glsp-tool.ts [hunk 13]
```diff
@@ -14,21 +14,28 @@
...
 export abstract class BaseGLSPTool implements GLSPTool {
     @inject(TYPES.IFeedbackActionDispatcher) protected feedbackDispatcher: IFeedbackActionDispatcher;
     @inject(TYPES.IActionDispatcher) protected actionDispatcher: IActionDispatcher;
-    @inject(MouseTool) protected mouseTool: MouseTool;
-    @inject(KeyTool) protected keyTool: KeyTool;
+    @inject(GLSPMouseTool) protected mouseTool: GLSPMouseTool;
+    @inject(GLSPKeyTool) protected keyTool: GLSPKeyTool;
     @inject(EditorContextService) protected readonly editorContext: EditorContextService;
 
+    protected onDisable = new DisposableCollection();
```
[the remark this answers is not in the document]
Comment (martin-fleck-at) [reply packages/client/src/features/tools/base-glsp-tool.ts:32]: `toDispose` is a bit too generic for my taste but how about `toDisposeOnDisable`? This is a similar naming convention that is used in Theia where they have `toDisposeOnDetach` on widgets.

File: packages/client/src/base/feedback/feedback-action-dispatcher.ts [hunk 14]
```diff
@@ -59,41 +60,44 @@ export interface IFeedbackActionDispatcher {
...
-                result.push(key);
+        this.registeredFeedback.forEach((actions, emitter) => {
+            if (actions.includes(action)) {
+                result.push(emitter);
             }
         });
         return result;
     }
+
+    protected dispatchFeedback(actions: Action[], feedbackEmitter: IFeedbackEmitter): void {
```
[the remark this answers is not in the document]
Comment (martin-fleck-at) [reply packages/client/src/base/feedback/feedback-action-dispatcher.ts:97]: Good catch, I'll change it.

File: packages/client/src/container-modules.ts (hunk 2 above)
Comment (martin-fleck-at) [reply packages/client/src/container-modules.ts:55]: Ah yes, I'm always happy to provide more consistency!

File: packages/client/src/features/tool-feedback/di.config.ts (hunk 7 above)
Comment (martin-fleck-at) [reply packages/client/src/features/tool-feedback/di.config.ts:19]: I moved this around many times so I'm glad to take any good suggestion and that is definitely one! I'll move it.

File: packages/client/src/features/tools/change-bounds/change-bounds-tool-feedback.ts (hunk 9 above)
Comment (martin-fleck-at) [reply packages/client/src/features/tools/change-bounds/change-bounds-tool-feedback.ts:283]: Ha! Good catch!

File: packages/client/src/features/tools/base-glsp-tool.ts (hunk 8 above)
Comment (martin-fleck-at) [reply packages/client/src/features/tools/base-glsp-tool.ts:54]: Very nice! You are right, I'll fix it!

File: packages/client/src/features/tools/change-bounds/change-bounds-tool.ts (hunk 1 above)
Comment (martin-fleck-at) [reply packages/client/src/features/tools/change-bounds/change-bounds-tool.ts:170]: Yeah, it was a bad merge, I'll restore the old behavior.

File: packages/client/src/features/tools/change-bounds/change-bounds-tool.ts (hunk 10 above)
Comment (martin-fleck-at) [reply packages/client/src/features/tools/change-bounds/change-bounds-tool.ts:315]: Yes!

**Comment by martin-fleck-at** [comment]
@planger @tortmayr Thank you very much for the fast feedback! I addressed all concerns. I do feel there is still something ab it off with the change bounds tool when you resize an element and then click somewhere else but that is also true on master as far as I can see. Maybe one of you has any pointers or it is a non-issue after all ;-)

**Comment by martin-fleck-at** [comment]
@tortmayr Thank you! I added another commit that should fix all the behavior now.

**Comment by tortmayr** [comment]
Thanks for the fast fix. 
Works almost perfectly now but I noticed on other issue:
Clicking on the tool palette after an resize (whilest the resize handles are still shown) results 
in unintended behavior. Seems like there is a feedback action dispatched that resizes the element to its previous state. On the next server update (e.g. by creating a node) the actual size is restored.
Peek 2023-07-13 13-04

**Comment by martin-fleck-at** [comment]
@tortmayr Good catch! With the latest fix everything should work as expected.
